axelor-purchase/src/main/java/com/axelor/apps/purchase/web/PurchaseOrderLineController.java

```diff
@@ -339,4 +339,38 @@ public void supplierPartnerDomain(ActionRequest request, ActionResponse response
 
     response.setAttr("supplierPartner", "domain", domain);
   }
+
+  public void productDomain(ActionRequest request, ActionResponse response) {
+
+    if (request.getContext().get("isFilterOnSupplier") == null) {
+      return;
+    }
+    boolean isFilterOnSupplier = (boolean) request.getContext().get("isFilterOnSupplier");
+
+    String domain = "";
+    if (isFilterOnSupplier) {
+      PurchaseOrder purchaseOrder = request.getContext().getParent().asType(PurchaseOrder.class);
+
+      if (purchaseOrder != null
+          && !purchaseOrder.getSupplierPartner().getSupplierCatalogList().isEmpty()) {
+        domain =
+            "self.id IN "
+                + purchaseOrder
+                    .getSupplierPartner()
+                    .getSupplierCatalogList()
+                    .stream()
+                    .map(it -> it.getProduct().getId())
+                    .collect(Collectors.toList())
+                    .toString()
+                    .replace('[', '(')
+                    .replace(']', ')');
+      } else {
+        domain = "self.id = 0";
+      }
+    } else {
+      domain =
+          "self.isModel = false and self.expense = false and (self.endDate = null or self.endDate > :__date__) and self.purchasable = true";
```

Comment:
the domain "self.isModel = false and self.expense = false and (self.endDate = null or self.endDate > :__date__) and self.purchasable = true" should be also applied when usFilterOnSupplier is true